[dev-first] API level interceptors gives error

**Description:**
When I define API level interceptors it gives an error while build command



**Suggested Assignees:**


**Affected Product Version:**
3.0.0-alpha
**OS, DB, other environment details and versions:**    

**Steps to reproduce:**
1. Define a definition yaml like for swagger petstore yaml
`apis:
    /petstore/v1:
        title: Swagger Petstore
        version: 1.0.0
        production_endpoint:
            type: 'http'
            urls: 
                - 'https://petstore.swagger.io/v2'
        sandbox_endpoint:
            type: 'http'
            urls: 
                - 'https://sand.petstore.ballerina.io/'
        request_interceptor: validateRequest
        response_interceptor: validateResponse
        resources:
            /pet/findByStatus:
                get:
                    production_endpoint:
                        type: 'http'
                        urls: 
                            - 'http://www.mocky.io/v2/5cbd4d1d2f0000e70a16cc0e'
                    throttle_policy: 3PerMin
                    
        

        `
2. Then it gives the following error when building 
`[2019-04-23 11:02:11,227] ERROR {org.wso2.apimgt.gateway.cli.cmd.Main} - Error while reading the definitions.yaml 
org.wso2.apimgt.gateway.cli.exception.CLIRuntimeException: Error while reading the definitions.yaml
	at org.wso2.apimgt.gateway.cli.utils.MgwDefinitionUtils.setMgwDefinition(MgwDefinitionUtils.java:47)
	at org.wso2.apimgt.gateway.cli.cmd.BuildCmd.execute(BuildCmd.java:102)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.wso2.apimgt.gateway.cli.cmd.Main.main(Main.java:51)
Caused by: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "request_interceptor" (class org.wso2.apimgt.gateway.cli.model.mgwdefinition.MgwAPIDefinition), not marked as ignorable (9 known properties: "security", "resources", "request-interceptor", "title", "response-interceptor", "production_endpoint", "sandbox_endpoint", "version", "cors"])
 at [Source: (File); line: 13, column: 45] (through reference chain: org.wso2.apimgt.gateway.cli.model.mgwdefinition.MgwRootDefinition["apis"]->org.wso2.apimgt.gateway.cli.model.mgwdefinition.MgwAPIsDefinition["/petstore/v1"]->org.wso2.apimgt.gateway.cli.model.mgwdefinition.MgwAPIDefinition["request_interceptor"])
	at com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException.from(UnrecognizedPropertyException.java:60)
	at com.fasterxml.jackson.databind.DeserializationContext.handleUnknownProperty(DeserializationContext.java:822)
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer.handleUnknownProperty(StdDeserializer.java:1151)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownProperty(BeanDeserializerBase.java:1567)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownVanilla(BeanDeserializerBase.java:1545)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:293)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.deser.SettableAnyProperty.deserialize(SettableAnyProperty.java:154)
	at com.fasterxml.jackson.databind.deser.SettableAnyProperty.deserializeAndSet(SettableAnyProperty.java:134)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownVanilla(BeanDeserializerBase.java:1539)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:293)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:127)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:287)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4001)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2890)
	at org.wso2.apimgt.gateway.cli.utils.MgwDefinitionUtils.setMgwDefinition(MgwDefinitionUtils.java:44)
	... 3 more`


**Related Issues:**
<!-- Any related issues such as sub tasks, issues reported in other repositories (e.g component repositories), similar problems, etc. -->

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/AddAPICmd.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/AddAPICmd.java
@@ -154,7 +154,7 @@ public class AddAPICmd implements GatewayLauncherCmd {
         String grpc;
         isOverwriteRequired = false;
         projectName = GatewayCmdUtils.buildProjectName(projectName);
-        RouteUtils.setRoutesConfigPath(GatewayCmdUtils.getProjectRoutesConfFilePath(projectName));
+        RouteUtils.setRoutesConfigPath(GatewayCmdUtils.getProjectMgwDefinitionFilePath(projectName));
 
         if (!new File(File.separator + projectName).exists()) {
             throw GatewayCmdUtils.createUsageException("Project name `" + projectName

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/AddRouteCmd.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/AddRouteCmd.java
@@ -37,7 +37,7 @@ public class AddRouteCmd implements GatewayLauncherCmd {
     public void execute() {
 
         projectName = GatewayCmdUtils.buildProjectName(projectName);
-        RouteUtils.setRoutesConfigPath(GatewayCmdUtils.getProjectRoutesConfFilePath(projectName));
+        RouteUtils.setRoutesConfigPath(GatewayCmdUtils.getProjectMgwDefinitionFilePath(projectName));
 
         if (resource_id == null || resource_id.isEmpty()) {
             if ((resource_id = GatewayCmdUtils.promptForTextInput(outStream, "Enter Resource ID: "))

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/BuildCmd.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/BuildCmd.java
@@ -99,7 +99,7 @@ public class BuildCmd implements GatewayLauncherCmd {
             try{
                 String toolkitConfigPath = GatewayCmdUtils.getMainConfigLocation();
                 init(projectName, toolkitConfigPath);
-                MgwDefinitionUtils.setMgwDefinition(projectLocation + "/definition.yaml");
+                MgwDefinitionUtils.configureMgwDefinition(projectName);
                 CodeGenerator codeGenerator = new CodeGenerator();
                 ThrottlePolicyGenerator policyGenerator = new ThrottlePolicyGenerator();
                 boolean changesDetected;

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/FunctionCmd.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/FunctionCmd.java
@@ -67,7 +67,7 @@ public class FunctionCmd implements GatewayLauncherCmd {
     public void execute() {
 
         String projectName = GatewayCmdUtils.getSingleArgument(mainArgs);
-        RouteUtils.setRoutesConfigPath(GatewayCmdUtils.getProjectRoutesConfFilePath(projectName));
+        RouteUtils.setRoutesConfigPath(GatewayCmdUtils.getProjectMgwDefinitionFilePath(projectName));
         File projectLocation = new File(GatewayCmdUtils.getProjectDirectoryPath(projectName)
                 + File.separator +"src");
 

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/ListAPIsCmd.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/ListAPIsCmd.java
@@ -44,7 +44,7 @@ public class ListAPIsCmd implements GatewayLauncherCmd {
     @Override
     public void execute() {
         projectName = GatewayCmdUtils.buildProjectName(projectName);
-        RouteUtils.setRoutesConfigPath(GatewayCmdUtils.getProjectRoutesConfFilePath(projectName));
+        RouteUtils.setRoutesConfigPath(GatewayCmdUtils.getProjectMgwDefinitionFilePath(projectName));
         printAPIDetailsInTable(RouteUtils.listApis());
     }
 

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/ListResourcesCmd.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/ListResourcesCmd.java
@@ -44,7 +44,7 @@ public class ListResourcesCmd implements GatewayLauncherCmd{
     @Override
     public void execute() {
         projectName = GatewayCmdUtils.buildProjectName(projectName);
-        RouteUtils.setRoutesConfigPath(GatewayCmdUtils.getProjectRoutesConfFilePath(projectName));
+        RouteUtils.setRoutesConfigPath(GatewayCmdUtils.getProjectMgwDefinitionFilePath(projectName));
 
         if (mainArgs == null) {
             printResourceDetailsForSingleAPI(OpenAPICodegenUtils.getAllResources(projectName));

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/UpdateRouteCmd.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/UpdateRouteCmd.java
@@ -39,7 +39,7 @@ public class UpdateRouteCmd implements GatewayLauncherCmd{
         if(projectName == null || projectName.isEmpty()){
             throw new CLIRuntimeException("Project name is not provided.");
         }
-        RouteUtils.setRoutesConfigPath(GatewayCmdUtils.getProjectRoutesConfFilePath(projectName));
+        RouteUtils.setRoutesConfigPath(GatewayCmdUtils.getProjectMgwDefinitionFilePath(projectName));
 
         if((apiId == null || apiId.isEmpty()) && (resourceId == null || resourceId.isEmpty())){
             throw new CLIRuntimeException("Error: API Id or resource id is not provided.");

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/constants/GatewayCliConstants.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/constants/GatewayCliConstants.java
@@ -58,6 +58,7 @@ public class GatewayCliConstants {
     public static final String EXTENSION_BALX = ".balx";
     public static final String EXTENSION_ZIP = ".zip";
     public static final String EXTENSION_JAR = ".jar";
+    public static final String EXTENSION_BAL = ".bal";
     public static final String GW_TARGET_DIST = "distribution";
     public static final String GW_DIST_PREFIX = "micro-gw-";
     public static final String GW_DIST_BIN = "bin";

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/mgwdefinition/MgwAPIDefinition.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/mgwdefinition/MgwAPIDefinition.java
@@ -125,8 +125,8 @@ public class MgwAPIDefinition {
      *
      * @return true if the API information is used inside
      */
-    public boolean getIsUsed(){
-        return isUsed;
+    public boolean getIsDefinitionUsed(){
+        return isDefinitionUsed;
     }
 
     /**
@@ -134,7 +134,7 @@ public class MgwAPIDefinition {
      *
      * @param flag true if the API information is used
      */
-    public void setIsUsed(boolean flag){
-        isUsed = flag;
+    public void setIsDefinitionUsed(boolean flag){
+        isDefinitionUsed = flag;
     }
 }

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/mgwdefinition/MgwAPIsDefinition.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/mgwdefinition/MgwAPIsDefinition.java
@@ -52,7 +52,7 @@ public class MgwAPIsDefinition {
         for (Map.Entry<String, MgwAPIDefinition> apiEntry : apis.entrySet()) {
             if (apiEntry.getValue().getTitle().equals(apiName) && apiEntry.getValue().getVersion().equals(apiVersion)) {
                 //because if this method is called, it means that API information is utilized.
-                apiEntry.getValue().setIsUsed(true);
+                apiEntry.getValue().setIsDefinitionUsed(true);
                 return apiEntry.getKey();
             }
         }

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java
@@ -54,6 +54,8 @@ public class BallerinaOperation implements BallerinaOpenAPIObject<BallerinaOpera
     private MgwEndpointConfigDTO epConfig;
     private String requestInterceptor;
     private String responseInterceptor;
+    private String apiRequestInterceptor;
+    private String apiResponseInterceptor;
 
     // Not static since handlebars can't see static variables
     private final List<String> allMethods =
@@ -207,5 +209,21 @@ public class BallerinaOperation implements BallerinaOpenAPIObject<BallerinaOpera
 
     public void setResponseInterceptor(String responseInterceptor) {
         this.responseInterceptor = responseInterceptor;
+    }
+
+    public String getApiRequestInterceptor() {
+        return apiRequestInterceptor;
+    }
+
+    public void setApiRequestInterceptor(String requestInterceptor) {
+        this.apiRequestInterceptor = requestInterceptor;
+    }
+
+    public String getApiResponseInterceptor() {
+        return apiResponseInterceptor;
+    }
+
+    public void setApiResponseInterceptor(String responseInterceptor) {
+        this.apiResponseInterceptor = responseInterceptor;
     }
 }

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaService.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaService.java
@@ -162,40 +162,53 @@ public class BallerinaService implements BallerinaOpenAPIObject<BallerinaService
         for (Map.Entry<String, PathItem> path : pathList.entrySet()) {
             BallerinaPath balPath = new BallerinaPath().buildContext(path.getValue(), this.api);
             balPath.getOperations().forEach(operation -> {
-                //todo: check the behavior if miscellaneous characters appear, if so remove the if condition
-                if (operation.getValue().getOperationId() == null) {
-                    // set the ballerina function name as {http_method}{UUID} ex : get_2345_sdfd_4324_dfds
-                    String operationId = operation.getKey() + "_" + UUID.randomUUID().toString().replaceAll("-", "_");
-                    operation.getValue().setOperationId(operationId);
-
-                }
+                // set the ballerina function name as {http_method}{UUID} ex : get_2345_sdfd_4324_dfds
+                String operationId = operation.getKey() + "_" + UUID.randomUUID().toString().replaceAll("-", "_");
+                operation.getValue().setOperationId(operationId);
+                //if it is the developer first approach
                 if (isDevFirst) {
                     String basePath = MgwDefinitionUtils.getBasePath(openAPI.getInfo().getTitle(),
                             openAPI.getInfo().getVersion());
                     //to add resource level endpoint configuration
                     MgwEndpointConfigDTO epConfig = MgwDefinitionUtils.getResourceEpConfigForCodegen(basePath,
                             path.getKey(), operation.getKey());
-                    if(epConfig != null){
+                    if (epConfig != null) {
                         operation.getValue().setEpConfigDTO(epConfig);
                     }
                     //todo: need to validate the existence of those functions
                     //to add request interceptor
                     String requestInterceptor = MgwDefinitionUtils.getRequestInterceptor(basePath, path.getKey(),
                             operation.getKey());
-                    if(requestInterceptor != null){
+                    if (requestInterceptor != null) {
                         operation.getValue().setRequestInterceptor(requestInterceptor);
                     }
                     //to add response interceptor
                     String responseInterceptor = MgwDefinitionUtils.getResponseInterceptor(basePath, path.getKey(),
                             operation.getKey());
-                    if(responseInterceptor != null){
+                    if (responseInterceptor != null) {
                         operation.getValue().setResponseInterceptor(responseInterceptor);
                     }
                     //to add throttle policy
                     String throttle_policy = MgwDefinitionUtils.getThrottlePolicy(basePath, path.getKey(),
                             operation.getKey());
-                    if(throttle_policy != null){
+                    if (throttle_policy != null) {
                         operation.getValue().setResourceTier(throttle_policy);
+                    }
+
+                    //to add API level request interceptor
+                    String apiRequestInterceptor = MgwDefinitionUtils.getApiRequestInterceptor(basePath);
+                    if (apiRequestInterceptor != null) {
+                        //if user specify the same interceptor function in both api level and resource level ignore
+                        // api level interceptor
+                        if (!apiRequestInterceptor.equals(requestInterceptor)) {
+                            operation.getValue().setApiRequestInterceptor(apiRequestInterceptor);
+                        }
+                    }
+
+                    //to add API level response interceptor
+                    String apiResponseInterceptor = MgwDefinitionUtils.getApiResponseInterceptor(basePath);
+                    if (apiResponseInterceptor != null) {
+                        operation.getValue().setApiResponseInterceptor(apiResponseInterceptor);
                     }
                 }
             });

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/GatewayCmdUtils.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/GatewayCmdUtils.java
@@ -732,12 +732,12 @@ public class GatewayCmdUtils {
     }
 
     /**
-     * Returns the path to the routes configuration file (routes.yaml)
+     * Returns the path to the microgateway definition file (definition.yaml).
      *
      * @param projectName name of the project
      * @return path to the client-cert-metadata.yaml for a defined version of an API
      */
-    public static String getProjectRoutesConfFilePath(String projectName) {
+    public static String getProjectMgwDefinitionFilePath(String projectName) {
         return getProjectDirectoryPath(projectName) + File.separator + GatewayCliConstants.PROJECT_DEFINITION_FILE;
     }
 
@@ -1266,10 +1266,8 @@ public class GatewayCmdUtils {
      * @param apiId md5 hash value of apiName:apiVersion
      * @return path to the swagger for a defined version of an API
      */
-    public static String getProjectSwaggerFilePath(String projectName, String apiId){
+    public static String getProjectSwaggerFilePath(String projectName, String apiId) {
         return getProjectAPIFilesDirectoryPath(projectName) + File.separator + apiId + File.separator +
                 GatewayCliConstants.API_SWAGGER;
     }
-
-
 }

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/MgwDefinitionUtils.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/MgwDefinitionUtils.java
@@ -215,31 +215,14 @@ public class MgwDefinitionUtils {
             LOGGER.error(errorMsg);
             throw new CLIRuntimeException(errorMsg);
         }
-     * To validate the provided definition.yaml follows the yaml syntax
-     *
-     * @param filePath file path to definition.yaml
-     */
-    private static void validateYamlSyntax(String filePath) {
-        File file = new File(filePath);
-        //to check the existence of definitions.yaml file
-        if (!file.exists()) {
-            throw GatewayCmdUtils.createValidationException("'definition.yaml' file does not exists.", LOGGER);
-        }
-        try {
-            OBJECT_MAPPER_YAML.readTree(file);
-            //if the provided definitions.yaml file does not follow the yaml syntax
-        } catch (IOException e) {
-            throw GatewayCmdUtils.createValidationException("'definitions.yaml file cannot be parsed as yaml document.",
-                    LOGGER);
-        }
     }
 
     /**
      * To find out the api information which is not used for code generation but included in the definitions.yaml
      */
-    public static void FindNotUsedAPIInformation() {
+    public static void FindNotUsedAPIInformation(){
         rootDefinition.getApis().getApisMap().forEach((k, v) -> {
-            if (!v.getIsUsed()) {
+            if (!v.getIsDefinitionUsed()) {
                 String msg = "API '" + v.getTitle() + "' version: '" + v.getVersion() + "' is not used but " +
                         "added to the " + GatewayCliConstants.PROJECT_DEFINITION_FILE + ".";
                 LOGGER.warn(msg);

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/MgwDefinitionUtils.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/MgwDefinitionUtils.java
@@ -19,6 +19,9 @@ package org.wso2.apimgt.gateway.cli.utils;
 
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.apimgt.gateway.cli.constants.GatewayCliConstants;
 import org.wso2.apimgt.gateway.cli.exception.CLIRuntimeException;
 import org.wso2.apimgt.gateway.cli.model.mgwcodegen.MgwEndpointConfigDTO;
 import org.wso2.apimgt.gateway.cli.model.mgwdefinition.MgwRootDefinition;
@@ -35,14 +38,16 @@ import java.io.IOException;
 public class MgwDefinitionUtils {
 
     private static final ObjectMapper OBJECT_MAPPER_YAML = new ObjectMapper(new YAMLFactory());
+    private static final Logger LOGGER = LoggerFactory.getLogger(MgwDefinitionUtils.class);
     //private static String routesConfigPath;
     private static MgwRootDefinition rootDefinition;
+    private static String projectName;
 
-    public static void setMgwDefinition(String path) {
+    public static void configureMgwDefinition(String project) {
+        projectName = project;
         try {
-            if (rootDefinition == null) {
-                rootDefinition = OBJECT_MAPPER_YAML.readValue(new File(path), MgwRootDefinition.class);
-            }
+            String definitionFilePath = GatewayCmdUtils.getProjectMgwDefinitionFilePath(project);
+            rootDefinition = OBJECT_MAPPER_YAML.readValue(new File(definitionFilePath), MgwRootDefinition.class);
         } catch (IOException e) {
             throw new CLIRuntimeException("Error while reading the definitions.yaml", e);
         }
@@ -69,42 +74,70 @@ public class MgwDefinitionUtils {
         return rootDefinition.getApis().getApiFromBasepath(basePath).getCorsConfiguration();
     }
 
-    public static MgwEndpointConfigDTO getResourceEpConfigForCodegen(String basePath, String path, String method) {
-        if(!isResourceAvailable(basePath, path, method)){
+    public static MgwEndpointConfigDTO getResourceEpConfigForCodegen(String basePath, String path, String operation) {
+        if(!isResourceAvailable(basePath, path, operation)){
             return null;
         }
         EndpointListRouteDTO prodList = rootDefinition.getApis().getApiFromBasepath(basePath).getPathsDefinition().
-                getMgwResource(path).getEndpointListDefinition(method).getProdEndpointList();
+                getMgwResource(path).getEndpointListDefinition(operation).getProdEndpointList();
         EndpointListRouteDTO sandList = rootDefinition.getApis().getApiFromBasepath(basePath).getPathsDefinition().
-                getMgwResource(path).getEndpointListDefinition(method).getSandEndpointList();
+                getMgwResource(path).getEndpointListDefinition(operation).getSandEndpointList();
         return RouteUtils.convertToMgwServiceMap(prodList,sandList);
     }
 
-    public static String getRequestInterceptor(String basePath, String path, String method){
-        if(!isResourceAvailable(basePath, path, method)){
+    /**
+     * Get the request interceptor for a given resource in an API.
+     *
+     * @param basePath basePath of the API
+     * @param path path variable of the resource
+     * @param operation operation of the resource
+     * @return  request interceptor of the resource if available, otherwise null.
+     */
+    public static String getRequestInterceptor(String basePath, String path, String operation) {
+        if (!isResourceAvailable(basePath, path, operation)) {
+            return null;
+        }
+        String interceptor = rootDefinition.getApis().getApiFromBasepath(basePath).getPathsDefinition().getMgwResource(path).
+                getEndpointListDefinition(operation).getRequestInterceptor();
+        validateInterceptorAvailability(interceptor, basePath, path, operation);
+        return interceptor;
+    }
+
+    /**
+     * Get the response interceptor for a given resource in an API.
+     *
+     * @param basePath basePath of the API
+     * @param path path variable of the resource
+     * @param operation operation of the resource
+     * @return  response interceptor of the resource if available, otherwise null.
+     */
+    public static String getResponseInterceptor(String basePath, String path, String operation) {
+        if (!isResourceAvailable(basePath, path, operation)) {
+            return null;
+        }
+        String interceptor = rootDefinition.getApis().getApiFromBasepath(basePath).getPathsDefinition().getMgwResource(path).
+                getEndpointListDefinition(operation).getResponseInterceptor();
+        validateInterceptorAvailability(interceptor, basePath, path, operation);
+        return interceptor;
+    }
+
+    public static String getThrottlePolicy(String basePath, String path, String operation){
+        if(!isResourceAvailable(basePath, path, operation)){
             return null;
         }
         return rootDefinition.getApis().getApiFromBasepath(basePath).getPathsDefinition().getMgwResource(path).
-                getEndpointListDefinition(method).getRequestInterceptor();
+                getEndpointListDefinition(operation).getThrottlePolicy();
     }
 
-    public static String getResponseInterceptor(String basePath, String path, String method){
-        if(!isResourceAvailable(basePath, path, method)){
-            return null;
-        }
-        return rootDefinition.getApis().getApiFromBasepath(basePath).getPathsDefinition().getMgwResource(path).
-                getEndpointListDefinition(method).getResponseInterceptor();
-    }
-
-    public static String getThrottlePolicy(String basePath, String path, String method){
-        if(!isResourceAvailable(basePath, path, method)){
-            return null;
-        }
-        return rootDefinition.getApis().getApiFromBasepath(basePath).getPathsDefinition().getMgwResource(path).
-                getEndpointListDefinition(method).getThrottlePolicy();
-    }
-
-    private static boolean isResourceAvailable(String basePath, String path, String method){
+    /**
+     * Check if the given resource is available in the definition.yaml.
+     *
+     * @param basePath basePath of the API
+     * @param path path of the resource
+     * @param operation operation of the resource
+     * @return true if the resource is available
+     */
+    private static boolean isResourceAvailable(String basePath, String path, String operation){
         if(rootDefinition.getApis().getApiFromBasepath(basePath) == null){
             return false;
         }
@@ -112,29 +145,62 @@ public class MgwDefinitionUtils {
             return false;
         }
         if(rootDefinition.getApis().getApiFromBasepath(basePath).getPathsDefinition().getMgwResource(path)
-                .getEndpointListDefinition(method) == null){
+                .getEndpointListDefinition(operation) == null){
             return false;
         }
         return true;
     }
 
     /**
-     * get the API-level request interceptor of an API.
+     * Get the API-level request interceptor of an API.
      *
      * @param basePath basePath of the API
-     * @return  API response request function name
+     * @return API response request function name
      */
     public static String getApiRequestInterceptor(String basePath){
-        return rootDefinition.getApis().getApiFromBasepath(basePath).getRequestInterceptor();
+        String interceptor = rootDefinition.getApis().getApiFromBasepath(basePath).getRequestInterceptor();
+        validateInterceptorAvailability(interceptor, basePath, null, null);
+        return interceptor;
     }
 
     /**
-     * get the API-level response interceptor of an API.
+     * Get the API-level response interceptor of an API.
      *
      * @param basePath basePath of the API
      * @return  API response interceptor function name
      */
     public static String getApiResponseInterceptor(String basePath){
-        return rootDefinition.getApis().getApiFromBasepath(basePath).getResponseInterceptor();
+        String interceptor = rootDefinition.getApis().getApiFromBasepath(basePath).getResponseInterceptor();
+        validateInterceptorAvailability(interceptor, basePath, null, null);
+        return interceptor;
+    }
+
+    /**
+     * Validate the existence of the interceptor inside interceptors directory.
+     * Throws an runtime error if the interceptor is not found.
+     * if the provided interceptor name is null, 'null' will be returned.
+     *
+     * @param interceptorName name of the interceptor
+     * @param basePath basePath
+     * @param path path of the resource (if interceptor is Api level keep null)
+     * @param operation operation of the resource (if interceptor is Api level keep null)
+     */
+    private static void validateInterceptorAvailability(String interceptorName, String basePath, String path,
+                                                        String operation) {
+        if (interceptorName == null) {
+            return;
+        }
+        File file = new File(GatewayCmdUtils.getProjectInterceptorsDirectoryPath(projectName) +
+                File.separator + interceptorName + GatewayCliConstants.EXTENSION_BAL);
+        if (!file.exists()) {
+            String errorMsg = "The interceptor '" + interceptorName + "' mentioned under basePath:'" + basePath + "' ";
+            //if the interceptor is resource level
+            if (path != null && operation != null) {
+                errorMsg += "path:'" + path + "' operation:'" + operation + "' ";
+            }
+            errorMsg += "is not available in the " + GatewayCliConstants.PROJECT_DEFINITION_FILE + ".";
+            LOGGER.error(errorMsg);
+            throw new CLIRuntimeException(errorMsg);
+        }
     }
 }

--- a/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/MgwDefinitionUtils.java
+++ b/components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/MgwDefinitionUtils.java
@@ -117,4 +117,24 @@ public class MgwDefinitionUtils {
         }
         return true;
     }
+
+    /**
+     * get the API-level request interceptor of an API.
+     *
+     * @param basePath basePath of the API
+     * @return  API response request function name
+     */
+    public static String getApiRequestInterceptor(String basePath){
+        return rootDefinition.getApis().getApiFromBasepath(basePath).getRequestInterceptor();
+    }
+
+    /**
+     * get the API-level response interceptor of an API.
+     *
+     * @param basePath basePath of the API
+     * @return  API response interceptor function name
+     */
+    public static String getApiResponseInterceptor(String basePath){
+        return rootDefinition.getApis().getApiFromBasepath(basePath).getResponseInterceptor();
+    }
 }

--- a/components/micro-gateway-cli/src/main/resources/templates/service.mustache
+++ b/components/micro-gateway-cli/src/main/resources/templates/service.mustache
@@ -99,7 +99,7 @@ service {{cut qualifiedServiceName " "}} on {{#api.transport}}{{#equals . "http"
     @gateway:RateLimit{policy : "{{resourceTier}}"}
     resource function {{operationId}} (http:Caller outboundEp, http:Request req) {
         handleExpectHeaderFor{{cut qualifiedServiceName " "}}(outboundEp, req);
-
+    {{#if apiRequestInterceptor}}{{apiRequestInterceptor}} (outboundEp, req);{{/if}}
     {{#if requestInterceptor}}{{requestInterceptor}} (outboundEp, req);{{/if}}
     {{#if api.isDefaultVersion}}
     string urlPostfix = untaint req.rawPath.replace("{{basepath}}","");
@@ -188,6 +188,7 @@ service {{cut qualifiedServiceName " "}} on {{#api.transport}}{{#equals . "http"
 
         if(clientResponse is http:Response) {
             {{#if responseInterceptor}}{{responseInterceptor}} (outboundEp, clientResponse);{{/if}}
+            {{#if apiResponseInterceptor}}{{apiResponseInterceptor}} (outboundEp, clientResponse);{{/if}}
             var outboundResult = outboundEp->respond(clientResponse);
             if (outboundResult is error) {
                 log:printError("Error when sending response", err = outboundResult);

